src/Microsoft.ComponentDetection.Detectors/rust/RustCliDetector.cs

```diff
@@ -0,0 +1,134 @@
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="RustCliDetector"/> class.
+    /// </summary>
+    /// <param name="componentStreamEnumerableFactory">The component stream enumerable factory.</param>
+    /// <param name="walkerFactory">The walker factory.</param>
+    /// <param name="cliService">The command line invocation service.</param>
+    /// <param name="logger">The logger.</param>
+    public RustCliDetector(
+        IComponentStreamEnumerableFactory componentStreamEnumerableFactory,
+        IObservableDirectoryWalkerFactory walkerFactory,
+        ICommandLineInvocationService cliService,
+        ILogger<RustCliDetector> logger)
+    {
+        this.ComponentStreamEnumerableFactory = componentStreamEnumerableFactory;
+        this.Scanner = walkerFactory;
+        this.cliService = cliService;
+        this.Logger = logger;
+    }
+
+    /// <inheritdoc />
+    public override string Id => "RustCli";
+
+    /// <inheritdoc />
+    public override IEnumerable<string> Categories { get; } = new[] { "Rust" };
+
+    /// <inheritdoc />
+    public override IEnumerable<ComponentType> SupportedComponentTypes => new[] { ComponentType.Cargo };
+
+    /// <inheritdoc />
+    public override int Version => 1;
+
+    /// <inheritdoc />
+    public override IList<string> SearchPatterns { get; } = new[] { "Cargo.toml" };
+
+    /// <inheritdoc />
+    protected override async Task OnFileFoundAsync(ProcessRequest processRequest, IDictionary<string, string> detectorArgs)
+    {
+        var componentStream = processRequest.ComponentStream;
+
+        try
+        {
+            if (!await this.cliService.CanCommandBeLocatedAsync("cargo", null))
+            {
+                this.Logger.LogWarning("Could not locate cargo command. Skipping Rust CLI detection");
+                return;
+            }
+
+            var cliResult = await this.cliService.ExecuteCommandAsync(
+                "cargo",
+                null,
+                "metadata",
+                "--manifest-path",
+                componentStream.Location,
+                "--format-version=1",
+                "--locked");
+
+            var metadata = CargoMetadata.FromJson(cliResult.StdOut);
+            var graph = BuildGraph(metadata);
+            var root = metadata.Resolve.Root;
+
+            this.TraverseAndRecordComponents(processRequest.SingleFileComponentRecorder, componentStream.Location, graph, root, null, null);
+        }
+        catch (Exception e)
+        {
+            this.Logger.LogWarning(e, "Failed to parse Cargo.toml file: {Location}", processRequest.ComponentStream.Location);
+        }
+    }
+
+    private static Dictionary<string, Node> BuildGraph(CargoMetadata cargoMetadata) => cargoMetadata.Resolve.Nodes.ToDictionary(x => x.Id);
+
+    private static (string Name, string Version) ParseNameAndVersion(string nameAndVersion)
+    {
+        var parts = nameAndVersion.Split(' ');
+        return (parts[0], parts[1]);
```

Comment:
Check how many parts we have here? And throw or log a warning if it's not exactly 2.

Reply:
The cargo format mandates that the format is `<crate-name> <version> (<registry-type>+<registry-url>)`